### PR TITLE
Add self-update command test

### PR DESCRIPTION
**Pull request by julienfalque**

### CODE REVIEW

File: src/Console/Command/SelfUpdateCommand.php [hunk 1]
```diff
@@ -123,36 +136,29 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        $remoteFilename = $this->toolInfo->getPharDownloadUri($remoteTag);
 
-        try {
-            $copyResult = @copy($remoteFilename, $tempFilename);
-            if (false === $copyResult) {
-                $output->writeln(sprintf('<error>Unable to download new version %s from the server.</error>', $remoteTag));
-
-                return 1;
-            }
+        $copyResult = @copy($remoteFilename, $tempFilename);
```
Comment (SpacePossum) [inline src/Console/Command/SelfUpdateCommand.php]: no need for variable introduction, it is only used one time in the `if` below

File: tests/TextDiffTest.php [hunk 2]
```diff
@@ -36,7 +36,9 @@
      */
     public function testDiffReportingDecorated($expected, $format, $isDecorated)
     {
-        $command = new FixCommand();
+        $command = new FixCommand(
+            $this->prophesize('PhpCsFixer\ToolInfo')->reveal()
```
Comment (SpacePossum) [inline tests/TextDiffTest.php]: does this test need a `prophesize` version of `ToolInfo`?

File: src/Console/SelfUpdate/NewVersionChecker.php [hunk 3]
```diff
@@ -19,7 +19,7 @@
 /**
  * @internal
  */
-final class NewVersionChecker
+class NewVersionChecker
```
Comment (SpacePossum) [inline src/Console/SelfUpdate/NewVersionChecker.php]: this is done for `prophesize` onyl?

File: src/Console/Command/SelfUpdateCommand.php [hunk 4]
```diff
@@ -75,21 +90,19 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+            $latestVersion = $this->versionChecker->getLatestVersion();
+            $latestVersionOfCurrentMajor = $this->versionChecker->getLatestVersionOfMajor($currentMajor);
         } catch (\Exception $exception) {
             $output->writeln(sprintf(
                 '<error>Unable to determine newest version: %s</error>',
                 $exception->getMessage()
             ));
 
-            return 0;
+            return 1;
```
Comment (SpacePossum) [inline src/Console/Command/SelfUpdateCommand.php]: good catch :+1: maybe this should go lower versions as well?

File: tests/TextDiffTest.php (hunk 2 above)
Comment (julienfalque) [reply tests/TextDiffTest.php]: Not sure, I'll have a look.

File: src/Console/Command/SelfUpdateCommand.php (hunk 4 above)
Comment (julienfalque) [reply src/Console/Command/SelfUpdateCommand.php]: Target branch is already 2.2 ;)

File: src/Console/SelfUpdate/NewVersionChecker.php (hunk 3 above)
Comment (julienfalque) [reply src/Console/SelfUpdate/NewVersionChecker.php]: Yes. This is exactly the reason of my comment on [#3183](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3183#discussion_r147670555).

Comment (keradus) [reply src/Console/SelfUpdate/NewVersionChecker.php]: please bring final and create interface.

File: src/Console/Command/SelfUpdateCommand.php (hunk 4 above)
Comment (keradus) [reply src/Console/Command/SelfUpdateCommand.php]: this bugfix deserves dedicated PR that gonna be merged sooner:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3196

File: src/PharChecker.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer;
+
+/**
+ * @internal
+ */
+class PharChecker
```
Comment (keradus) [inline src/PharChecker.php]: shall be final

**Comment by keradus** [comment]
big one, I will need deeper look.
can we somehow split it ?

File: src/Console/SelfUpdate/NewVersionChecker.php (hunk 3 above)
Comment (julienfalque) [reply src/Console/SelfUpdate/NewVersionChecker.php]: TBH I don't see the point in creating an interface for testing purposes only.

Comment (keradus) [reply src/Console/SelfUpdate/NewVersionChecker.php]: `final` prevents partial mocks, which are extra bad approach imo.
i like to be defensive.

Comment (julienfalque) [reply src/Console/SelfUpdate/NewVersionChecker.php]: What do you mean partial mocks? Mocks with some methods defined in the test and some not defined? If so, using an interface does not prevent this. Also, when using Prophecy, as soon as a mock has at least one defined method, all methods used in the tested code must be defined as well. Calling an undefined method will throw an exception, which should prevent the issue.

File: tests/TextDiffTest.php (hunk 2 above)
Comment (julienfalque) [reply tests/TextDiffTest.php]: Removed unneeded dummies.

File: src/Console/Command/SelfUpdateCommand.php (hunk 1 above)
Comment (julienfalque) [reply src/Console/Command/SelfUpdateCommand.php]: Variable removed.

File: src/Console/SelfUpdate/NewVersionChecker.php (hunk 3 above)
Comment (keradus) [reply src/Console/SelfUpdate/NewVersionChecker.php]: with built-in mock system of PHPUnit, it's possible to mock real class with using mocked one method but not another. final class, will forbid that totally.

Comment (julienfalque) [reply src/Console/SelfUpdate/NewVersionChecker.php]: We don't use PHPUnit mocks but Prophecy anyway so this still looks pointless to me.

Comment (keradus) [reply src/Console/SelfUpdate/NewVersionChecker.php]: final or not is a general rule. using final is being defensive. partial mocks were example, not the only reason.

**Comment by julienfalque** [comment]
Updated to keep final classes final and use interfaces for mocks instead.

@keradus I think some small parts could be extracted, but most of the changes are directly tied to adding the `SelfUpdateCommand` test, so not worth it IMO.

File: src/Console/Command/FixCommand.php [hunk 6]
```diff
@@ -68,14 +69,17 @@
      */
     private $defaultConfig;
 
-    public function __construct()
+    private $toolInfo;
```
Comment (keradus) [inline src/Console/Command/FixCommand.php]: typehinting is missing

File: src/Console/Application.php [hunk 7]
```diff
@@ -33,16 +37,24 @@
 {
     const VERSION = '2.2.10';
 
+    private $toolInfo;
```
Comment (keradus) [inline src/Console/Application.php]: typehinting is missing

File: src/Console/Command/SelfUpdateCommand.php [hunk 8]
```diff
@@ -33,6 +33,24 @@
 {
     const COMMAND_NAME = 'self-update';
 
+    private $versionChecker;
+
+    private $toolInfo;
+
+    private $pharChecker;
```
Comment (keradus) [inline src/Console/Command/SelfUpdateCommand.php]: typehintings are missing

File: src/Console/ConfigurationResolver.php [hunk 9]
```diff
@@ -104,6 +104,8 @@
      */
     private $configFinderIsOverridden;
 
+    private $toolInfo;
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: same

**Comment by keradus** [comment]
1.3k lines on diff, not sure if "not worth it" indeed

**Comment by julienfalque** [comment]
See #3246.

**Comment by keradus** [comment]
#3246 got merged. please merge/rebase with 2.2

**Comment by julienfalque** [comment]
Rebased.

File: src/PharCheckerInterface.php [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace PhpCsFixer;
+
+/**
+ * @internal
+ */
+interface PharCheckerInterface
+{
+    /**
+     * @param string $filename
```
Comment (keradus) [inline src/PharCheckerInterface.php]: what is the output ?

Comment (julienfalque) [reply src/PharCheckerInterface.php]: Updated.

**Comment by keradus** [comment]
Thank you @julienfalque.

**Pull request closed without merging**